src/charm.py

```diff
@@ -0,0 +1,178 @@
+    CredentialsGoneEvent,
+    S3Requirer,
+)
+from lightkube.core.client import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.resources.core_v1 import ServiceAccount
+from ops.charm import CharmBase, ConfigChangedEvent, InstallEvent
+from ops.main import main
+from ops.model import StatusBase
+
+from config import SparkConfigurationHubConfig
+from constants import (
+    CONF_FILE_PATH,
+    PEER,
+    S3_INTEGRATOR_REL,
+)
+from models import S3ConnectionInfo, Status
+from utils import WithLogging
+
+
+class SparkConfigurationHubCharm(CharmBase, WithLogging):
+    """Charm the service."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.framework.observe(self.on.install, self._update_event)
+        self.framework.observe(self.on.update_status, self._update_event)
+        self.framework.observe(self.on.install, self._on_install)
+        self.s3_requirer = S3Requirer(self, S3_INTEGRATOR_REL)
+        self.framework.observe(
+            self.s3_requirer.on.credentials_changed, self._on_s3_credential_changed
+        )
+        self.framework.observe(self.s3_requirer.on.credentials_gone, self._on_s3_credential_gone)
+
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.client = Client(field_manager=self.model.app.name)
+
+    @property
+    def peers(self):
+        """Fetch the peer relation."""
+        return self.model.get_relation(PEER)
+
+    def set_peer_data(self, key: str, data: Optional[str]) -> None:
+        """Set peer data."""
+        if data:
+            self.peers.data[self.app][key] = data
+        else:
+            del self.peers.data[self.app][key]
+
+    def get_peer_data(self, key: str) -> Optional[str]:
+        """Get peer data."""
+        if not self.peers:
+            return None
+        data = self.peers.data[self.app].get(key, None)
+        return data if data else None
+
+    @property
+    def k8s_permissions(self) -> bool:
+        """Check if the charm has the permission to inspect other namespaces."""
+        try:
+            _ = self.client.list(ServiceAccount, namespace="*")
+            return True
+        except ApiError:
+            return False
+
+    @property
+    def s3_connection_info(self) -> Optional[S3ConnectionInfo]:
+        """Parse a S3ConnectionInfo object from relation data."""
+        if not self.s3_requirer.relations:
+            return None
+
+        raw = self.s3_requirer.get_s3_connection_info()
+
+        return S3ConnectionInfo(
+            **{key.replace("-", "_"): value for key, value in raw.items() if key != "data"}
+        )
+
+    def get_status(
+        self,
+        s3: Optional[S3ConnectionInfo],
+    ) -> StatusBase:
+        """Compute and return the status of the charm."""
+        if not self.k8s_permissions:
+            return Status.MISSING_PERMISSIONS.value
+
+        if not s3:
+            return Status.MISSING_S3_RELATION.value
```

Comment:
**todo, question** But in this case, the configuration hub can do its job without s3. Just in this case, we don't add the s3 part to the configuration, right?